crates/re_log_encoding/src/decoder/stream.rs

```diff
@@ -0,0 +1,281 @@
+                if let Some(header) = self.chunks.try_read(FileHeader::SIZE)? {
+                    // header contains version and compression options
+                    self.compression = read_options(header)?.compression;
+
+                    // we might have data left in the current chunk,
+                    // immediately try to read length of the next message
+                    self.state = State::MessageHeader;
+                    return self.try_read();
+                }
+            }
+            State::MessageHeader => {
+                if let Some(mut len) = self.chunks.try_read(MessageHeader::SIZE)? {
+                    let header = MessageHeader::decode(&mut len)?;
+                    self.state = State::Message(header);
+                    // we might have data left in the current chunk,
+                    // immediately try to read the message content
+                    return self.try_read();
+                }
+            }
+            State::Message(header) => {
+                if let Some(bytes) = self.chunks.try_read(header.compressed as usize)? {
+                    let bytes = match self.compression {
+                        Compression::Off => bytes,
+                        Compression::LZ4 => {
+                            self.uncompressed.resize(header.uncompressed as usize, 0);
+                            lz4_flex::block::decompress_into(bytes, &mut self.uncompressed)
+                                .map_err(DecodeError::Lz4)?;
+                            &self.uncompressed
+                        }
+                    };
+
+                    // read the message from the uncompressed bytes
+                    let message = rmp_serde::from_slice(bytes).map_err(DecodeError::MsgPack)?;
+
+                    self.state = State::MessageHeader;
+                    return Ok(Some(message));
+                }
+            }
+        }
+
+        Ok(None)
+    }
+}
+
+impl Default for StreamDecoder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+type Chunk = Cursor<Vec<u8>>;
+
+struct ChunkBuffer {
+    /// Any incoming chunks are queued until they are emptied
+    queue: VecDeque<Chunk>,
+
+    /// When `try_read` is called and we don't have enough bytes yet,
+    /// we store whatever we do have in this buffer
+    buffer: Vec<u8>,
+
+    /// The cursor points to the end of the used range in `buffer`
+    cursor: usize,
+}
+
+impl ChunkBuffer {
+    fn new() -> Self {
+        Self {
+            queue: VecDeque::with_capacity(16),
+            buffer: Vec::with_capacity(1024),
+            cursor: 0,
+        }
+    }
+
+    fn push(&mut self, chunk: Vec<u8>) {
+        self.queue.push_back(Chunk::new(chunk));
+    }
+
+    /// Attempt to read `n` bytes out of the queued chunks.
+    ///
+    /// Returns `Ok(None)` if there is not enough data to return a slice of `n` bytes.
+    fn try_read(&mut self, n: usize) -> Result<Option<&[u8]>, DecodeError> {
```

Comment:
True, it's reading from `Cursor<Vec<u8>>`, and if that's empty it will just repeatedly return `Ok(0)` (at which point the chunk will be removed), so I'll replace this with an `expect` and make `try_read` infallible